gamemodes/terrortown/entities/entities/ttt_logic_role.lua

```diff
@@ -24,35 +26,46 @@ function ENT:KeyValue(key, value)
 			value = _G[value] or value
 		end
 
-		self.Role = tonumber(value)
+		self.checkingRole = tonumber(value)
 
-		if not self.Role then
+		if not self.checkingRole then
 			ErrorNoHalt("ttt_logic_role: bad value for Role key, not a number\n")
 
-			self.Role = ROLE_NONE
+			self.checkingRole = ROLE_NONE
 		end
 	end
 end
 
 ---
--- @param string name
--- @param Entity|Player activator
--- @return[default=true] boolean
--- @realm shared
-function ENT:AcceptInput(name, activator)
+-- Called when another entity fires an event to this entity.
+-- @param string name The name of the input that was triggered
+-- @param Entity|Player activator The initial cause for the input getting triggered (e.g. the player who pushed a button)
+-- @param Entity caller The entity that directly triggered the input (e.g. the button that was pushed)
+-- @param string data The data passed
+-- @return[default=true] boolean Return true if the default action should be supressed
+-- @realm server
+function ENT:AcceptInput(name, activator, caller, data)
 	if name == "TestActivator" then
-		if IsValid(activator) and activator:IsPlayer() then
-			local activator_role = (GetRoundState() == ROUND_PREP) and ROLE_INNOCENT or activator:GetBaseRole()
+		if not IsValid(activator) or not activator:IsPlayer() then return end
+
+		---
+		-- @realm server
+		local role, team = hook.Run("TTT2ModifyLogicRoleCheck", activator, self, activator, caller, data)
+		local activatorRole = roles.GetByIndex(role):GetBaseRole()
+		local activatorTeam = (GetRoundState() == ROUND_PREP) and TEAM_INNOCENT or team
 
-			if self.Role == ROLE_NONE or self.Role == activator_role then
-				Dev(2, activator, "passed logic_role test of", self:GetName())
+		if self.checkingRole == ROLE_TRAITOR and util.IsEvilTeam(activatorTeam)
+			or self.checkingRole == ROLE_INNOCENT and not util.IsEvilTeam(activatorTeam)
```

Comment:
This should explicitly check for team innocent imo. As the convar for counting evil teams would also affect this. Or remove this special case completely

Reply:
yes, you're absolutely right on this one!

Reply:
REmoving is no option here, as we check for innocent role what detectives definitely arent, but should be included.
This check makes sure, that in the classic way without the convar everything is inno, thats not a traitor and in the modern way makes sure only the innocent team is allowed.

So I'd say it should stay like this.

Reply:
Since this is a role check entity i dont really think innocent checks should be special. Would you mind explaining the usecase behind this team stuff and special handling in detail?

Reply:
The maps use different approaches to check for traitor. Here are some examples in the "classic" way:
1) Waterworld: Checks for Innocent, so every other role including the detective is invalid/a "traitor"
2) Minecraft: Counts all Traitors, so only traitors are included as valid
3) Minecraft: Checks for Traitor, so everyone else beside traitors are invalid
4) Innocentmotel: Checks for Detective, only he is valid

So our idea was to get those cases, that are used on maps valid in two ways.
The Classic case means, that really only traitors are counted as traitor and not inno.
The New Case with the Convar activated means, that only inno teams are counted as innocent and not traitor

Reply:
I think I see the problem @saibotk is describing here. Think about the following scenario: The convar is set to `false` and the player is `TEAM_NECROMANCER`.

https://github.com/TTT-2/TTT2/pull/809/files#diff-96dc1dc56ce17268ec8dd91f20b2ccfcf7747bd07ce4f11c40e01e32fc089c3bR457-R477

This function then returns `false`, which means this function passes a necromancer as an innocent. Is this the behavior we want in this case? Honestly I'm not 100% sure, it might be ok.

Reply:
I know, but thats why we implemented this convar in the first place. The old system didnt know about other roles, but they handled everything via traitor, inno and dete. Therefore we decided that a check for inno, which I've only seen in one place is actually a check for not traitor. So that old cases are still valid with the convar set to false. In my opinion I would just get rid of the convar and make it modern all the way, but this would be my idea of implementing it. If we decide for innocents we only use the modern way kinda seems wrong.

Reply:
I don't think the existence of the convar changes anything really. The way it was handled previously was that the check just failed, now it doesn't. Tbh both are wrong, because the maps aren't designed for aything but T, dete and inno. Therefore I'd say our implementation here is alright, because it isn't better or worse for this specific edge case, just different.

Reply:
I just think that we should deal with this one way.
Convar false: only traitor is "traitor" and everything else is "innocent"
Convar true: Only Team Inno is "innocent" and everything else is traitor

Reply:
This should have been dealt with by now because I removed the convar.